Sources/Swift/HangTracker.swift

```diff
@@ -0,0 +1,216 @@
+    let endTime: TimeInterval
+}
+
+#if SENTRY_TEST || SENTRY_TEST_CI || DEBUG
+protocol HangTracker {
+    // The callback must be called on a background thread, because the main thread is blocked
+    func addLateRunLoopObserver(handler: @escaping (UUID, TimeInterval) -> Void) -> UUID
+    
+    func removeLateRunLoopObserver(id: UUID)
+
+    // The callback is always called on the main thread
+    func addFinishedRunLoopObserver(handler: @escaping (RunLoopIteration) -> Void) -> UUID
+    
+    func removeFinishedRunLoopObserver(id: UUID)
+}
+protocol RunLoopObserver { }
+
+extension DefaultHangTracker: HangTracker { }
+extension CFRunLoopObserver: RunLoopObserver { }
+#else
+typealias HangTracker = DefaultHangTracker<CFRunLoopObserver>
+typealias RunLoopObserver = CFRunLoopObserver
+#endif
+
+typealias CreateObserverFunc<T> = (_ allocator: CFAllocator?, _ activities: CFOptionFlags, _ repeats: Bool, _ order: CFIndex, _ block: ((T?, CFRunLoopActivity) -> Void)?) -> T?
+typealias AddObserverFunc<T> = (_ rl: CFRunLoop?, _ observer: T?, _ mode: CFRunLoopMode?) -> Void
+typealias RemoveObserverFunc<T> = (_ rl: CFRunLoop?, _ observer: T?, _ mode: CFRunLoopMode?) -> Void
+
+final class DefaultHangTracker<T: RunLoopObserver> {
+
+    init(
+        dateProvider: SentryCurrentDateProvider,
+        createObserver: @escaping CreateObserverFunc<T>,
+        addObserver: @escaping AddObserverFunc<T>,
+        removeObserver: @escaping RemoveObserverFunc<T>,
+        queue: DispatchQueue = DispatchQueue(label: "io.sentry.runloop-observer-checker")
+    ) {
+        self.dateProvider = dateProvider
+        self.createObserver = createObserver
+        self.addObserver = addObserver
+        self.removeObserver = removeObserver
+        self.queue = queue
+#if canImport(UIKit) && !SENTRY_NO_UIKIT && (!swift(>=5.9) || !os(visionOS)) && !os(watchOS)
+        var maxFPS = 60.0
+        if #available(iOS 13.0, tvOS 13.0, *) {
+            let window = UIApplication.shared.connectedScenes.flatMap { ($0 as? UIWindowScene)?.windows ?? [] }.first { $0.isKeyWindow }
+            maxFPS = Double(window?.screen.maximumFramesPerSecond ?? 60)
+        } else {
+            maxFPS = Double(UIScreen.main.maximumFramesPerSecond)
+        }
+#else
+        let maxFPS: Double = 60.0
+        #endif
+        let expectedFrameDuration = 1.0 / maxFPS
+        hangNotifyThreshold = expectedFrameDuration * 1.5
+    }
+    
+    func addLateRunLoopObserver(handler: @escaping (UUID, TimeInterval) -> Void) -> UUID {
+        let id = UUID()
+        queue.async { [weak self] in
+            self?.lateRunLoop[id] = handler
+            DispatchQueue.main.async {
+                self?.startIfNecessary()
+            }
+        }
+        return id
+    }
+    
+    func removeLateRunLoopObserver(id: UUID) {
+        queue.async { [weak self] in
+            guard let self else { return }
+            lateRunLoop.removeValue(forKey: id)
+            if lateRunLoop.isEmpty {
```

Comment:
`h`: Seer already mentioned it, I believe the two nested `async` statements could allow a race-condition